src/socket/set.rs

```diff
@@ -174,6 +178,142 @@ impl<'a, 'b: 'a, 'c: 'a + 'b> Set<'a, 'b, 'c> {
     }
 }
 
+impl<'a, 'b: 'a, 'c: 'a + 'b> PacketFilter for Set<'a, 'b, 'c> {
+    fn process_udp(&mut self, ip_repr: &IpRepr, udp_repr: &UdpRepr) -> Result<()> {
+        for mut udp_socket in self.iter_mut().filter_map(UdpSocket::downcast) {
+            if !udp_socket.accepts(&ip_repr, &udp_repr) { continue }
+
+            match udp_socket.process(&ip_repr, &udp_repr) {
+                // The packet is valid and handled by socket.
+                Ok(()) => return Ok(()),
+                // The packet is malformed, or the socket buffer is full.
+                Err(e) => return Err(e)
+            }
+        }
+        Err(Error::Dropped)
+    }
+
+    fn process_tcp<'frame>(&mut self, timestamp: u64, ip_repr: &IpRepr, tcp_repr: &TcpRepr<'frame>) ->
+        Result<Option<(IpRepr, TcpRepr<'static>)>>
+    {
+        for mut tcp_socket in self.iter_mut().filter_map(TcpSocket::downcast) {
+            if !tcp_socket.accepts(&ip_repr, &tcp_repr) { continue }
+
+            return tcp_socket.process(timestamp, &ip_repr, &tcp_repr);
+        }
+        Err(Error::Dropped)
+    }
+
+    fn process_icmpv4(&mut self, ip_repr: &IpRepr, icmp_repr: &Icmpv4Repr, checksum_caps: &ChecksumCapabilities) -> Result<()> {
+        let mut handled = false;
+        for mut icmp_socket in self.iter_mut().filter_map(IcmpSocket::downcast) {
+            if !icmp_socket.accepts(&ip_repr, &icmp_repr, checksum_caps) { continue }
+
+            match icmp_socket.process(&ip_repr, icmp_repr, checksum_caps) {
+                // The packet is valid and handled by socket.
+                Ok(()) => handled = true,
+                // The socket buffer is full.
+                Err(Error::Exhausted) => (),
+                // ICMP sockets don't validate the packets in any way.
+                Err(_) => unreachable!(),
+            }
+        }
+        if handled { Ok(()) } else { Err(Error::Dropped) }
+    }
+
+    fn process_raw(&mut self, ip_repr: &IpRepr, ip_payload: &[u8], checksum_caps: &ChecksumCapabilities) -> Result<()> {
+        let mut handled = false;
+        for mut raw_socket in self.iter_mut().filter_map(RawSocket::downcast) {
+            if !raw_socket.accepts(&ip_repr) { continue }
+
+            match raw_socket.process(&ip_repr, ip_payload, &checksum_caps) {
+                // The packet is valid and handled by socket.
+                Ok(()) => handled = true,
+                // The socket buffer is full.
+                Err(Error::Exhausted) => (),
+                // Raw sockets don't validate the packets in any way.
+                Err(_) => unreachable!(),
+            }
+        }
+        if handled { Ok(()) } else { Err(Error::Dropped) }
```

Comment:
This is quite confusing. A packet not being accepted by any raw socket is not an error condition; this is how `process_raw` is different from e.g. `process_tcp`. Why not just use `Result<bool>` for the return value of this function? Then it'd just be `Ok(handled)`...

There are three different semantics here. A handler for raw sockets can:
  * take the packet and eat it (putting it into its own buffer), which counts as success;
  * ignore the packet and let the higher layer handle it, which also counts as success;
  * deliberately drop the packet (think of an application like a firewall).

Only the third one should result in an `Error::Dropped`, IMO.

Reply:
Makes sense. I'll change the return type to Result<bool>.